### ISSUE TITLE
Culling massive Blockwise graphs is very slow, not constant-time

### ISSUE THREAD

**Issue by gjoseph92** [dask/dask#8570]
**tl;dr: Could `get_all_external_keys` and `get_output_keys` be avoided in HLG culling?**

In some workflows, it can be desirable to create a dask Array/DataFrame structure representing some full-size, enormous dataset, then immediately use slicing to sub-select out only a tiny part of it, then work with that. Now that we can use Blockwise for IO (xref https://github.com/dask/dask/pull/7417), this is an especially appealing pattern, because it should be constant-time to construct the massive graph, since nothing has to be materialized.

I had hoped it would also be linear-time to cull this massive graph, but it appears currently that it's not.

Here's an example where I'm trying to create an xarray representing the [Landsat-8](https://planetarycomputer.microsoft.com/dataset/landsat-8-c2-l2) collection at full resolution over the entire continental US. This is a 10PB, 136-million-chunk array that involves 1.3 billion data loading tasks. Here I'm using https://github.com/gjoseph92/stackstac/pull/116 (so the data loading graph is fully blockwise) and https://github.com/dask/dask/pull/8560 (so we know `fuse_roots` isn't materializing the graph unnecessarily).

![Screen Shot 2022-01-14 at 10 37 54 AM](https://user-images.githubusercontent.com/3309802/149560244-8b8bec05-f3e5-4427-a079-3714677e77a7.png)

Then I'm sub-selecting a single chunk out of those 136 million. Based on what I know of the graph, this should cull down to 4 tasks.

![Screen Shot 2022-01-14 at 10 37 59 AM](https://user-images.githubusercontent.com/3309802/149560279-e40a3d55-aedf-449b-b931-a9f2fd2a1645.png)

<details><summary>Here's the HLG for reference</summary>

You can see the first layer is materialized with ~100,000 tasks, but the _big_, 100-million-task one is Blockwise. (https://github.com/dask/dask/issues/8497 sure would be nice here!)

![Screen Shot 2022-01-14 at 10 30 55 AM](https://user-images.githubusercontent.com/3309802/149559272-8f5e0f13-863e-49ac-8482-5876cd2b9617.png)

</details>

But when I try to optimize this graph, I see memory usage shoot up until it crashes the kernel on the 32GB machine. Interrupting the kernel after a few seconds makes it pretty clear what's going on: `HighLevelGraph.cull` is calling `get_all_external_keys`, which is forcing the generation of all 1.3 billion keys (or 136 million keys? not sure).

![Screen Shot 2022-01-14 at 10 23 19 AM](https://user-images.githubusercontent.com/3309802/149558357-d981f103-cea9-44bb-b44e-ed7070003b77.png)

Even if it didn't call `get_all_external_keys`, I see that `HighLevelGraph.cull` is still calling `get_output_keys` on every layer. For reference: https://github.com/dask/dask/blob/358a5e367eedc8f2963651071a72f43e0ac4f887/dask/highlevelgraph.py#L944-L970

Why is it necessary for HLG.cull to ask the layer for all its keys, intersect them itself, then pass that back into the layer? And why is it necessary for cull functions to take `all_hlg_keys`?

I had thought the interface would be simply: HLG.cull tells each layer the necessary output keys; the layer figures out the rest on its own. If it needs to generate all its keys internally and do that intersection, fine, but for layers that don't need to do this, shouldn't the optimization be available?

@rjzamora @madsbk why does culling work this way? Would it be possible to write Blockwise culling without this, in a way that's truly linear-time to only the number of final keys?

cc @ian-r-rose @TomAugspurger

**Comment by gjoseph92** [comment]
So would it be possible to rewrite the culling process to not require any layers to generate all their keys?

**Comment by martindurant** [comment]
Totally agree with @gjoseph92 , I was under the impression that there were two major benefits of HLGs in general:
- smaller transfer size, since we do something similar to using a generator/comprehension rather than a concrete set of keys and tasks
- that you can select/update from a large graph at no cost, effectively changing the start/end of the comprehension implied in the first bullet.
- (possibilities for higher-level optimisations)

**Comment by jakirkham** [comment]
I could be wrong about this, but given the proposed changes around HLG transfer ( https://github.com/dask/distributed/issues/5581 ) there is an opportunity to improve transfer size (if more work is still needed there)

**Comment by martindurant** [comment]
@jakirkham , that seems like an orthogonal discussion. In the case here, the HLG is expensive to construct before even starting to send it, even though in principle the selection should result in a small graph without having to iterate over the original. So it's point 2 in my list which is the surprising omission.

**Comment by timblakely** [comment]
Ran into this as well. Was working with a 700k x 270k x 2k volume and even `__setitem__`-ing a small 5x5x5 slice was quite slow. Digging into it showed that it was enumerating >400k blocks to cull down to the single one it needed.

Sounds like #9216 may help, but I'm not familiar enough with the inner working of `Layer`s to know for sure.

**Comment by d-v-b** [comment]
I'm also frequently running into this issue. In particular, slow slicing prevents dask arrays from being a performant backend for chunk-aware data visualization tools like neuroglancer.

**Comment by martindurant** [comment]
+1+1+1

Given that, conceptually, an array layer is like a dict comprehension
```
llg = {f"name_{keys}": (func, args) for keys in zip(axis_range0, axis_range1, ...)}
```
where how to make the args is layer-specific, it should be possible to cull by changing the start/stop/step of the axes ranges. We should never be materialising the full set of keys.

**Comment by gjoseph92** [comment]
Yup, agreed this definitely shouldn't be the behavior, and the complexity of the current design makes this hard to avoid. I don't know if there's any upcoming work planned around this @jrbourbeau @ian-r-rose @rjzamora?

**Comment by rjzamora** [comment]
> I don't know if there's any upcoming work planned around this @jrbourbeau @ian-r-rose @rjzamora?

Once serialization is simplified in https://github.com/dask/distributed/pull/6028 (or something else similar?), I'm planning to remove  the `Mapping` backend from `Layer` and remove both `Layer.cull` and `HighLevelGraph.cull` altogether.  My (strong-ish) opinion is that we should make it impossible to "accidentally" materialize a `Layer`, by not supporting `Mapping`-like attributes. Instead, the Layer should cull itself when it is explicitly materialized by the scheduler.  I think a simple but strict Layer API can go a long way.

**Comment by r-pastor** [comment]
Also ran into this problem recently while working on a web app that access a big archive of geospatial data.

A unidimensional simplified example of the issue can be run with the following:
```python
da.full((1e8), fill_value=np.nan, chunks=(10))[0].compute()  # This takes a few seconds to complete
```
As @gjoseph92 commented, the problem seems to be with the fact that all the keys of the graph are created when just one is actually needed.

If we assume that when there is a layer on the graph performing a selection such as the _getitem_ there is a lot of branches that can be pruned, there should be a way of optimizing the process by not creating such keys at all.

I have briefly looked into the code with that assumption in mind and seems like setting the `output_blocks` attribute of the `Blockwise` layer would make the `get_output_keys` call produce just the desired tasks. The value to set such attribute can be retrieved from the _getitem_ task dependency. 
In any case, my naive attempt of implementing it has broken a lot of tests. Furthermore I am not even sure if this is the correct track to solve the problem as the logic of the HLG cull and all the object involved is not straight forward to me.

Could some of the maintainers provide some guidance on how to fix this issue?
